File: it-agents/skills/code-architect/references/code-examples/cpp/Decorator.cpp
// Decorator — attach responsibilities by wrapping objects at runtime.
#include <iostream>
#include <memory>
#include <string>

// Component interface.
class Stream {
public:
    virtual ~Stream() = default;
    virtual std::string write(const std::string& data) const = 0;
};

// Concrete component.
class RawStream : public Stream {
public:
    std::string write(const std::string& data) const override { return data; }
};

// Base decorator wraps another Stream.
class StreamDecorator : public Stream {
protected:
    std::unique_ptr<Stream> inner;
public:
    explicit StreamDecorator(std::unique_ptr<Stream> s) : inner(std::move(s)) {}
};

class CompressDecorator : public StreamDecorator {
public:
    using StreamDecorator::StreamDecorator;
    std::string write(const std::string& data) const override {
        return "compressed(" + inner->write(data) + ")";
    }
};

class EncryptDecorator : public StreamDecorator {
public:
    using StreamDecorator::StreamDecorator;
    std::string write(const std::string& data) const override {
        return "encrypted(" + inner->write(data) + ")";
    }
};

int main() {
    std::unique_ptr<Stream> s = std::make_unique<RawStream>();
    s = std::make_unique<CompressDecorator>(std::move(s));
    s = std::make_unique<EncryptDecorator>(std::move(s));

    std::cout << s->write("payload") << "\n";
    return 0;
}

File: it-agents/skills/code-architect/references/code-examples/python/decorator.py
"""Decorator — attach responsibilities by wrapping objects at runtime."""

from abc import ABC, abstractmethod


class Stream(ABC):
    @abstractmethod
    def write(self, data: str) -> str:
        ...


class RawStream(Stream):
    def write(self, data: str) -> str:
        return data


class StreamDecorator(Stream):
    """Base wrapper holding a reference to the wrapped component."""

    def __init__(self, wrapped: Stream) -> None:
        self._wrapped = wrapped

    def write(self, data: str) -> str:
        return self._wrapped.write(data)


class UpperCaseDecorator(StreamDecorator):
    def write(self, data: str) -> str:
        return super().write(data.upper())


class BracketDecorator(StreamDecorator):
    def write(self, data: str) -> str:
        return super().write(f"[{data}]")


if __name__ == "__main__":
    stream = BracketDecorator(UpperCaseDecorator(RawStream()))
    print(stream.write("payload"))
